src/util/bip152.rs

```diff
@@ -0,0 +1,441 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! BIP152 Compact Blocks
+//!
+//! Implementation of compact blocks data structure and algorithms.
+//!
+
+use crate::prelude::*;
+
+use core::{convert, fmt, mem};
+use crate::io;
+#[cfg(feature = "std")] use std::error;
+
+use hashes::{sha256, siphash24, Hash};
+
+use crate::{BlockHash, Transaction, Block, BlockHeader};
+use crate::consensus::encode::{self, Decodable, Encodable, VarInt};
+use crate::util::endian;
+
+/// A BIP-152 error
+#[derive(Clone, PartialEq, Eq, Debug, Copy, PartialOrd,Ord, Hash)]
+pub enum Error {
+    /// An unknown version number was used.
+    UnknownVersion,
+    /// The prefill slice provided was invalid.
+    InvalidPrefill,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Error::UnknownVersion => write!(f, "an unknown version number was used"),
+            Error::InvalidPrefill => write!(f, "the prefill slice provided was invalid"),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl error::Error for Error {}
+
+/// A [PrefilledTransaction] structure is used in [HeaderAndShortIds] to
+/// provide a list of a few transactions explicitly.
+#[derive(PartialEq, Eq, Clone, Debug, PartialOrd, Ord, Hash)]
+pub struct PrefilledTransaction {
+    /// The index of the transaction in the block.
+    ///
+    /// This field is differentially encoded relative to the previous
+    /// prefilled transaction as described as follows:
+    ///
+    /// > Several uses of CompactSize below are "differentially encoded". For
+    /// > these, instead of using raw indexes, the number encoded is the
+    /// > difference between the current index and the previous index, minus one.
+    /// > For example, a first index of 0 implies a real index of 0, a second
+    /// > index of 0 thereafter refers to a real index of 1, etc.
+    pub idx: u16,
+    /// The actual transaction.
+    pub tx: Transaction,
+}
+
+impl convert::AsRef<Transaction> for PrefilledTransaction {
+    fn as_ref(&self) -> &Transaction {
+        &self.tx
+    }
+}
+
+impl Encodable for PrefilledTransaction {
+    #[inline]
+    fn consensus_encode<S: io::Write + ?Sized>(&self, mut s: &mut S) -> Result<usize, io::Error> {
+        Ok(VarInt(self.idx as u64).consensus_encode(&mut s)? + self.tx.consensus_encode(&mut s)?)
+    }
+}
+
+impl Decodable for PrefilledTransaction {
+    #[inline]
+    fn consensus_decode<D: io::Read + ?Sized>(mut d: &mut D) -> Result<PrefilledTransaction, encode::Error> {
+        let idx = VarInt::consensus_decode(&mut d)?.0;
+        if idx > u16::max_value() as u64 {
+            //TODO use TryInto instead when MSRV exceeds 1.34
```

Comment:
implement TODO since we are now on 1.41.1?

Reply:
Yep, this works
```rust
        let idx = VarInt::consensus_decode(&mut d)?.0;
        let idx = u16::try_from(idx).map_err(|_| encode::Error::ParseFailed("BIP152 prefilled tx index out of bounds"))?;
        let tx = Transaction::consensus_decode(&mut d)?;

        Ok(PrefilledTransaction { idx, tx })
```